torch/distributions/gumbel.py

```diff
@@ -0,0 +1,51 @@
+from numbers import Number
+
+import math
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all
+
+
+class Gumbel(Distribution):
+    r"""
+    Samples from a Gumbel Distribution.
+
+    Examples::
+
+        >>> m = Gumbel(torch.Tensor([1.0]), torch.Tensor([2.0]))
+        >>> m.sample()  # sample from Gumbel distribution with loc=1, scale=2
+         1.0124
+        [torch.FloatTensor of size 1]
+
+    Args:
+        loc (float or Tensor or Variable): Location parameter of the distribution
+        scale (float or Tensor or Variable): Scale parameter of the distribution
+    """
+    has_rsample = True
+    params = {'loc': constraints.real, 'scale': constraints.positive}
+    support = constraints.real
+
+    def __init__(self, loc, scale):
+        self.loc, self.scale = broadcast_all(loc, scale)
+        if isinstance(loc, Number) and isinstance(scale, Number):
+            batch_shape = torch.Size()
+        else:
+            batch_shape = self.scale.size()
+        super(Gumbel, self).__init__(batch_shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        uni_dist = self.scale.new(shape).uniform_()
+        # X ~ Uniform(0, 1)
+        # Y = loc - scale * ln (-ln (X)) ~ Gumbel(loc, scale)
+        return self.loc - self.scale * torch.log(-uni_dist.log())
```

Comment:
Can this produce `-inf` if `unform_` happens to sample exactly 0?

Reply:
Yes, you are right. In `aten/src/TH/THRandom.c` the uniform double is generated in the interval [0, 1). I will try to fix this ASAP.

Reply:
I have fixed it and the quantiles are looking good and consistent with the ones in `scipy.stats.gumbel_r`. Thank you for raising this issue.